src/runtime_src/core/edge/sw_em/generic_pcie_hal2/halapi.cxx

```diff
@@ -76,26 +76,31 @@ xclDeviceHandle xclOpen(unsigned deviceIndex, const char *logfileName, xclVerbos
   FeatureRomHeader fRomHeader;
   std::memset(&fRomHeader, 0, sizeof(FeatureRomHeader));
 
-  xclcpuemhal2::CpuemShim *handle = nullptr;
+  xclcpuemhal2::CpuemShim *shimObj = NULL;
+  void* handle;
+
   bool bDefaultDevice = false;
   std::map<unsigned int, xclcpuemhal2::CpuemShim*>::iterator it = xclcpuemhal2::devices.find(deviceIndex);
   if(it != xclcpuemhal2::devices.end())
   {
-    handle = (*it).second;
+    xclcpuemhal2::shimHandleMap[handle] = (*it).second;
```

Comment:
You can save shimHandle here and use it below instead of calling handle check everytime